plugins/inputs/kubernetes/kubernetes.go

```diff
@@ -70,15 +78,83 @@ func (k *Kubernetes) Init() error {
 	}
 	k.labelFilter = labelFilter
 
+	if k.URL == "" {
+		k.InsecureSkipVerify = true
+	}
+
 	return nil
 }
 
 // Gather collects kubernetes metrics from a given URL
 func (k *Kubernetes) Gather(acc telegraf.Accumulator) error {
-	acc.AddError(k.gatherSummary(k.URL, acc))
+	if k.URL != "" {
+		acc.AddError(k.gatherSummary(k.URL, acc))
+		return nil
+	}
+
+	var wg sync.WaitGroup
+	nodeBaseURLs, err := getNodeURLs(k.Log)
+	if err != nil {
+		return err
+	}
+
+	for _, url := range nodeBaseURLs {
+		wg.Add(1)
+		go func(url string) {
+			defer wg.Done()
+			acc.AddError(k.gatherSummary(url, acc))
+		}(url)
+	}
+	wg.Wait()
+
 	return nil
 }
 
+func getNodeURLs(log telegraf.Logger) ([]string, error) {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, err
+	}
+	client, err := kubernetes.NewForConfig(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	nodes, err := client.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+
+	nodeUrls := make([]string, 0, len(nodes.Items))
+	for _, n := range nodes.Items {
+		address := getNodeAddress(n)
+		if address == "" {
+			log.Warn("Unable to node addresses for Node '%s'", n.Name)
+			continue
+		}
+		nodeUrls = append(nodeUrls, "https://"+address+":10250")
+	}
+
+	return nodeUrls, nil
+}
+
+// Prefer internal addresses, if none found, use ExternalIP
+func getNodeAddress(node v1.Node) string {
+	extAddresses := make([]string, 0)
+
+	for _, addr := range node.Status.Addresses {
+		if addr.Type == v1.NodeInternalIP {
+			return addr.Address
+		}
+		extAddresses = append(extAddresses, addr.Address)
+	}
+
+	if len(extAddresses) > 0 {
+		return extAddresses[0]
```

Comment:
For multiple addresses would this be if a cluster had ipv6 and ipv4 configured? How safe is this assumption to grab the first one?

Reply:
Here I replicated behaviour of prometheus operator https://github.com/prometheus-operator/prometheus-operator/blob/ba18c613100f9693cacbd09d5735ac196eed68e1/pkg/prometheus/operator.go#L545-L558

we take first internall address, but if missing (I am not sure such setup is possible at all TBH), take first external address.